packages/graph/hash_graph/lib/graph/src/types/schema/combinator.rs

```diff
@@ -0,0 +1,157 @@
+use serde::{Deserialize, Serialize};
+
+use crate::types::schema::{array::TypedArray, ValidationError};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Optional<T> {
+    None {},
+    Some(T),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum ValueOrArray<T> {
+    Value(T),
+    Array(TypedArray<T>),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct OneOfRepr<T> {
+    one_of: Vec<T>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(try_from = "OneOfRepr<T>", rename_all = "camelCase")]
+pub struct OneOf<T> {
+    #[serde(flatten)]
+    repr: OneOfRepr<T>,
+}
+
+impl<T> OneOf<T> {
+    /// Creates a new `OneOf` without validating.
+    pub fn new_unchecked<U: Into<Vec<T>>>(one_of: U) -> Self {
+        Self {
+            repr: OneOfRepr {
+                one_of: one_of.into(),
+            },
+        }
+    }
+
+    /// Creates a new `OneOf` from the given vector.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if the object is not in a valid state.
+    pub fn new<U: Into<Vec<T>>>(one_of: U) -> Result<Self, ValidationError> {
+        let one_of = Self::new_unchecked(one_of);
+        one_of.validate()?;
+        Ok(one_of)
+    }
+
+    #[must_use]
+    pub fn one_of(&self) -> &[T] {
+        &self.repr.one_of
+    }
+
+    fn validate(&self) -> Result<(), ValidationError> {
```

Comment:
Why is this not `must_use`, surely this one is the one we want it on most? 😂

Reply:
Because `Result` is `#[must_use]`.

Reply:
Ah gotcha, it's implicit haha.

Cheers